### ISSUE TITLE
Compile in multiple stages when the number of source files in the transitive closure is much larger than the number of immediate dependencies

### ISSUE THREAD

**Issue by surajacharya** [sbt/sbt#322]
I work with a large project where there are few files that are depended on by many other files, which results in many files being recompiled when almost any files are edited to have API changes. Right now it looks like sbt starts from the files which have API changes and finds the transitive closure of the source files dependent on them, and compiles the full set. 

Would it still be correct to check if there's a large difference in size between the number of immediate dependents and the transitive closure, and if this were the case to compile just the immediate dependents, find the artifacts with API changes and then repeat the process of checking for immediate and transitive dependents?

**Comment by Blaisorblade** [comment]
You might be experiencing, in addition, issue #288. But let's ignore that for now, since your proposal has independent merit.

To me the suggested approach sounds safe and interesting. However, I found a (corner?) case where it is not beneficial, which is when mutual dependencies are involved. I do not know what actually SBT implements - I speculate that SBT uses your algorithm and is otherwise "sufficiently smart" (as smart as possible). After showing this problematic example, I suggest an improved variant.

When a changed file F1 both affects directly and depends on F2 (which is unchanged), an API change in F1 will cause both F1 and F2 to be recompiled again. Now suppose additionally that F2 affects and is affected by F3: after recompiling F1 and F2, if F2 has an API change, all three files will be recompiled. We compile F1 the first time, F1 and F2 the second, F1, F2 and F3 the 3rd, for a total of 1 + 2 + 3 = 6 compilations of a files. Extending the example to n files we perform a quadratic number of compilations in total (more precisely, n \* (n + 1) / 2).

My conclusion is that SBT should add, to the list L of files to recompile, also all transitive dependencies on which some file in L depends; to handle the example above, this processing step should be run in a fixpoint loop. This way, if more files need to be recompiled, they won't trigger recompilation of the original files.

More importantly, I conjecture (as a desirable goal) that each file should be recompiled at most twice (or if that's not possible, a constantly bounded number of times): once if recompilation is triggered directly for any reason, and at most one additional time if its recompilation triggered recompilation of some of its transitive dependencies.

**Comment by harrah** [comment]
@surajacharya Your understanding of what sbt does is correct.  As @Blaisorblade suggests, there are correctness issues involving cycles.

Another reason sbt has not had more than two steps (1. changed files 2. transitive dependencies of files with API changes) is that the overhead of starting up a new compiler instance is high.

Recently, experimental support for keeping the compiler resident was added.  This will help address the first problem.  Once it is more stable, I will probably look into doing invalidation in three steps instead of two as well as check to see the fraction of the sources being recompiled is above a certain threshold.  This is the kind of change that doesn't break binary compatibility, so it can wait for a 0.12.1 or .2 release.

**Comment by harrah** [comment]
@Blaisorblade With respect to your comment in #288, please feel free to add a wiki page.  The center of the incremental logic is in [compile/inc/Incremental.scala](https://github.com/harrah/xsbt/blob/0.13/compile/inc/Incremental.scala).  The [old page](https://code.google.com/p/simple-build-tool/wiki/ChangeDetectionAndTesting) for 0.7 and earlier may also be useful.

**Comment by Blaisorblade** [comment]
@harrah Thanks! I got started with a first draft, which is however still orphan on purpose (it contains XXX comments).
https://github.com/harrah/xsbt/wiki/Understanding-incremental-recompilation
Please feel free to take a look at it, especially at Section "How to take advantage of SBT heuristics" - I skimmed the implementation only briefly, hence some statements are only educated guesses from my year-long experience using SBT.

I focused on issues which were non-intuitive for me when I first noticed them using SBT, maybe because one could imagine in theory a smarter SBT handling them (whether it can be implemented or whether it would be faster is another issue). One could even consider some of these issues as bugs. For instance, I write that "Modifying the interface of a non-private method requires recompiling all clients, even if the method is not used", but one can imagine how to address that problem. Similarly, adding a method does not require recompiling all clients&mdash;a more precise approximation would be possible computing enough information, it seems just extremely hard.
Assume classes A1, ... An depend on class B and SBT tracks for each file (including A1 ..., An) which APIs it actually uses (including APIs from B), SBT could understand whether adding or removing a member to B in fact requires recompiling classes A1, ..., An - it could also understand which ones need to be recompiled. However, figuring out what is the needed API seems potentially a research project by itself: for instance, when adding method `B.foo`, we need to know whether client relied on the absence of `B.foo` because they used a pimped version of it. I expect that more such problems might show up. If pimped methods could overload available methods (they currently cannot, but I would hope for this to get fixed), the needed tracking would be much more complex.

**Comment by harrah** [comment]
@Blaisorblade Great!  I will try to provide more detailed feedback when I get a chance, but here are some quick things:
1. There are 45 scripted tests on incremental compilation here: https://github.com/harrah/xsbt/tree/0.13/sbt/src/sbt-test/source-dependencies.  new-cyclic and false-error are tests that currently fail.  They are examples of problems with a multi-stage approach that can get worse with more stages (such as only recompiling direct dependencies).
2. There are quite a few tickets in the Scala issues tracker (search for 'build manager') illustrating the difficulties of doing more refined invalidation.  Sealed classes are a good example.
3. sbt tracks source dependencies at the source level and binary dependencies at the class and file level.  In particular, it does not separate classes in the same source file (sealed classes again, for example).
4. Perhaps use code examples to illustrate problems with inferred return type.
5. Implicits are a problem, but the more general problem is actually naming and scopes.  With inheritance, there is the need to check overriding (either that it is as expected by an 'override' modifier or that it isn't unintentional overriding something).  The implementation details of traits require recompilation of descendant classes on most signature changes.  lazy vals until recently (2.10 trunk) required recompiling descendants.  Things like these should be described in various bug reports and test cases.

**Comment by asouza** [comment]
Hi, as we have discussed in https://github.com/harrah/xsbt/issues/288 , I am still experiencing the transitive compilation problem, even with the latest build of sbt. I'm using the 0.13 version.

**Comment by Blaisorblade** [comment]
@asouza: this problem is known to not be yet addressed. Mark Harrah declared that he'll try to address it after releasing 0.12.
@harrah:
You wrote that "3. sbt tracks source dependencies at the source level and binary dependencies at the class and file level. In particular, it does not separate classes in the same source file (sealed classes again, for example)." I tried incorporating this feedback, and I will later address point your point 4. (I just committed my changes).

I looked into issues with sealed classes and tried to extrapolate tracking of binary dependencies at the _file level_: I guess you might mean the source file from which each class originates, declared in the `SourceFile` attribute. This attribute is however declared by the VM specification to be optional, hence my guess (tracking the original source file) seems difficult to implement reliably.

It seems that for both cases, the tracking could be done at the level of the sealed class: if I modify the sealed hierarchy rooted at `sealed class/trait X`, dependencies of `X` should be recompiled. Of course, the algorithm must take into account that each class might be part of multiple sealed hierarchies. What am I missing?

The clarity of the new text is a bit lower, also because I am less sure on the issues themselves, but I will address that later.

**Comment by harrah** [comment]
The SourceFile attribute is not used for Scala classes.  This information can be easily [obtained from scalac](https://github.com/harrah/xsbt/blob/0.13/compile/interface/Analyzer.scala#L36).  For Java sources, it is used to determine the source file name when present, but there is a fallback that derives it from the class name.  In the Java case, it has to be guessed either way.  That logic is in [util/classfile/Analyze.scala](https://github.com/harrah/xsbt/blob/0.13/util/classfile/Analyze.scala#L40).

You might find it useful to look at the [Analysis type](http://harrah.github.com/xsbt/latest/api/index.html#sbt.inc.Analysis).  All of the information extracted from sources and classes is included there.  It may help clarify what relationships and timestamps are used.  The fundamental methods are the ones that return `Relation` instances.  The others are convenience wrappers around the relations.

For a source File, Analysis contains:
1. dependencies on source files in the current project or an upstream project it depends on
2. for binary dependencies (no source available), the name of the class and the jar file or class file providing that class
3.  the classes generated for the source file

As for trying to track dependencies at a finer granularity than the source API per source file, that is for a next generation incremental compiler.  The benefit is low compared to using a resident compiler or finer invalidation steps.  I expect the conceptual complexity of granular dependencies to be O(N^2) in the number of Scala features N compared to O(N) for the current source API approach.  So, I'm not convinced that kind of improvement should be considered before a working resident compiler and finer invalidation steps.

**Comment by harrah** [comment]
Implemented in 089b4e284c26601fda72b4b4cc0403865bcf3dae.  Heuristics on project size are not included at this time.
